File: 백준/Gold/14502. 연구소/연구소.py
from collections import deque
import copy


def dfs(wall,arr):
    global ans
    if wall == 3:
        ret = virus(arr)

        if ans < ret:
            ans = ret
        return

    for i in range(len(zero)):
        if used[i] == 1:
            return
        used[i] = 1
        backup = copy.deepcopy(arr)
        backup[zero[i][0]][zero[i][1]] = 1
        dfs(wall+1,backup)
        backup[zero[i][0]][zero[i][1]] = 0
        used[i] = 0

def virus(arr2):
    q = deque()
    visit = [[0]*M for _ in range(N)]
    for i in range(N):
        for j in range(M):
            if visit[i][j] == 1:
                continue
            if arr2[i][j] == 2:
                q.append((i,j))
                while q:
                    y,x = q.popleft()
                    for k in range(4):
                        dy = diy[k] + y
                        dx = dix[k] + x
                        if dy<0 or dy>=N or dx<0 or dx>=M:
                            continue
                        if arr2[dy][dx] == 1:
                            continue
                        if visit[dy][dx] == 0:
                            if arr2[dy][dx] == 0:
                                visit[dy][dx] = 1
                                arr2[dy][dx] = 2
                                q.append((dy,dx))
    cnt = 0
    for i in range(N):
        for j in range(M):
            if arr2[i][j] == 0:
                cnt += 1
    return cnt

N, M = map(int, input().split())
lst = [list(map(int,input().split())) for _ in range(N)]
zero = []
diy = [0,0,1,-1]
dix = [1,-1,0,0]
ans = 0
for i in range(N):
    for j in range(M):
        if lst[i][j] == 0:
            zero.append((i,j))
used = [0] * len(zero)

dfs(0,lst)
print(ans)

File: 백준/Gold/14502. 연구소/연구소.java
import java.util.*;

public class Main {
    static int N, M, ans = 0;
    static int[][] map;
    static List<int[]> zero = new ArrayList<>();
    static int[] dy = {0, 0, 1, -1};
    static int[] dx = {1, -1, 0, 0};
    static boolean[] used;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        map = new int[N][M];

        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++) {
                map[i][j] = sc.nextInt();
                if (map[i][j] == 0)
                    zero.add(new int[]{i, j});
            }

        used = new boolean[zero.size()];
        dfs(0, map);
        System.out.println(ans);
    }

    static void dfs(int wall, int[][] arr) {
        if (wall == 3) {
            int[][] copy = deepCopy(arr);
            ans = Math.max(ans, spreadVirus(copy));
            return;
        }

        for (int i = 0; i < zero.size(); i++) {
            if (used[i]) continue;
            used[i] = true;
            int[][] copy = deepCopy(arr);
            int y = zero.get(i)[0];
            int x = zero.get(i)[1];
            copy[y][x] = 1;
            dfs(wall + 1, copy);
            used[i] = false;
        }
    }

    static int spreadVirus(int[][] arr) {
        int[][] visited = new int[N][M];
        Queue<int[]> q = new ArrayDeque<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j] == 2 && visited[i][j] == 0) {
                    q.offer(new int[]{i, j});
                    while (!q.isEmpty()) {
                        int[] cur = q.poll();
                        int y = cur[0];
                        int x = cur[1];
                        for (int d = 0; d < 4; d++) {
                            int ny = y + dy[d];
                            int nx = x + dx[d];
                            if (ny < 0 || ny >= N || nx < 0 || nx >= M) continue;
                            if (arr[ny][nx] == 0 && visited[ny][nx] == 0) {
                                visited[ny][nx] = 1;
                                arr[ny][nx] = 2;
                                q.offer(new int[]{ny, nx});
                            }
                        }
                    }
                }
            }
        }

        int cnt = 0;
        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++)
                if (arr[i][j] == 0)
                    cnt++;
        return cnt;
    }

    static int[][] deepCopy(int[][] src) {
        int[][] copy = new int[N][M];
        for (int i = 0; i < N; i++)
            copy[i] = Arrays.copyOf(src[i], M);
        return copy;
    }
}
